quota.1
.TH QUOTA 1
.SH NAME
quota \- display disk usage and limits
.SH SYNOPSIS
.B quota
[
.B -F
.I format-name
] [
.B -guqvswi
] [
.BR -l \ |
[
.BR -QAm
]]
.br
.B quota
[
.B -F
.I format-name
] [
.B -qvswi
] [
.BR -l \ |
[
.BR -QAm
]]
.B -u
.IR user ...
.br
.B quota
[
.B -F
.I format-name
] [
.B -qvswi
] [
.BR -l \ |
[
.BR -QAm
]]
.B -g
.IR group ...
.br
.B quota
[
.B -F
.I format-name
] [
.B -qvswi
] [
.BR -l \ |
[
.BR -QAm
]]
.B -P
.IR project ...
.br
.B quota
[
.B -F
.I format-name
] [
.B -qvswugQm
]
.B -f
.IR filesystem ...
.SH DESCRIPTION
.B quota
displays users' disk usage and limits.
By default only the user quotas are printed. By default space usage and limits
are shown in kbytes (and are named blocks for historical reasons).
.PP
.B quota
reports the quotas of all the filesystems listed in
.BR /etc/mtab .
For filesystems that are NFS-mounted a call to the rpc.rquotad on
the server machine is performed to get the information.
.SH OPTIONS
.TP
.B -F, --format=\f2format-name\f1
Show quota for specified format (ie. don't perform format autodetection).
Possible format names are:
.B vfsold
Original quota format with 16-bit UIDs / GIDs,
.B vfsv0
Quota format with 32-bit UIDs / GIDs, 64-bit space usage, 32-bit inode usage and limits,
.B vfsv1
Quota format with 64-bit quota limits and usage,
.B rpc
(quota over NFS),
.B xfs
(quota on XFS filesystem)
.TP
.B -g, --group
Print group quotas for the group 
of which the user is a member.
The optional 
.B group
argument(s) restricts the display to the specified group(s).
.TP
.B -u, --user
flag is equivalent to the default.
.TP
.B -P, --project
Print project quotas for the specified project.
.TP
.B -v, --verbose
will display quotas on filesystems
where no storage is allocated.
.TP
.B -s, --human-readable[=\f2units\f1]
option will make
.BR quota (1)
try to choose units for showing limits, used space and used inodes. Units can
be also specified explicitely by an optional argument in format [
.BR kgt
],[
.BR kgt
] where the first character specifies space units and the second character
specifies inode units.
.TP
.B --always-resolve
Always try to translate user / group name to uid / gid even if the name
is composed of digits only.
.TP
.B -p, --raw-grace
When user is in grace period, report time in seconds since epoch when his grace
time runs out (or has run out). Field is '0' when no grace time is in effect.
This is especially useful when parsing output by a script.
.TP
.B -i, --no-autofs
ignore mountpoints mounted by automounter
.TP
.B -l, --local-only
report quotas only on local filesystems (ie. ignore NFS mounted filesystems).
.TP
.B -A, --all-nfs
report quotas for all NFS filesystems even if they report to be on the same
device.
.TP
.B -f, --filesystem-list
report quotas only for filesystems specified on command line.
.TP
.B --filesystem=\f2path\f1
report quotas only for filesystem
.BR path .
This option can be specified multiple types and quota will be reported for each
specified filesystem. Unlike command line option
.B -f
remaining command like arguments are still treated as user / group / project names
to report.
.TP
.B -m, --no-mixed-pathnames
Currently, pathnames of NFSv4 mountpoints are sent without leading slash in the path.
.BR rpc.rquotad
uses this to recognize NFSv4 mounts and properly prepend pseudoroot of NFS filesystem
to the path. If you specify this option, 
.BR quota 
will always send paths with a leading slash. This can be useful for legacy reasons but
be aware that quota over RPC will stop working if you are using new
.BR rpc.rquotad .
.TP
.B -q, --quiet
Print a more terse message,
containing only information
on filesystems where usage is over quota.
.TP
.B -Q, --quiet-refuse
Do not print error message if connection to
.BR rpc.rquotad
is refused (usually this happens when
.BR rpc.rquotad
is not running on the server).
.TP
.B -w, --no-wrap
Do not wrap the line if the device name is too long. This can be useful when parsing
the output of
.BR quota (1)
by a script.
.TP
.B --show-mntpoint
Show also mount point as a filesystem identification.
.TP
.B --hide-device
Do not show device name in a filesystem identification.
.LP
Specifying both
.B \-g
and
.B \-u
displays both the user quotas and the group quotas (for
the user).
.LP
Only the super-user may use the
.B \-u
flag and the optional
.B user
argument to view the limits of other users. Also viewing of project quota
usage and limits is limited to super-user only.
Non-super-users can use the
.B \-g
flag and optional
.B group
argument to view only the limits of groups of which they are members.
.LP
The
.B \-q
flag takes precedence over the
.B \-v
flag.
.SH DIAGNOSTICS
If
.B quota
exits with a non-zero status, one or more filesystems
are over quota.
.SH FILES
.PD 0
.TP 20
.B aquota.user " or " aquota.group
quota file at the filesystem root (version 2 quota, non-XFS filesystems)
.TP 20
.B quota.user " or " quota.group
quota file at the filesystem root (version 1 quota, non-XFS filesystems)
.TP
.B /etc/mtab
default filesystems
.PD
.SH SEE ALSO
.BR quotactl (2),
.BR fstab (5),
.BR edquota (8),
.BR quotacheck (8),
.BR quotaon (8),
.BR quota_nld (8),
.BR repquota (8),
.BR warnquota (8),
.BR setquota (8)


Changelog
Changes in quota-tools from 4.04 to 4.05
* Use getrpcbynumber() instead of getrpcbynumber_r() (Jan Kara)
* configure.ac: fix pkg_check_modules calls (Fabrice Fontaine)
* Include <fcntl.h> to get loff_t definition (Fabrice Fontaine)
* Add missing <sys/cdefs.h> include (Fabrice Fontaine)
* Use proper C99 integer types (Fabrice Fontaine)
* Disable quot tool (Jan Kara)
* Remove old disclaimers (Jan Kara)
* rpc.rquotad: Handle daemon(3) failure (Jan Kara)
* quota_nld: Check daemon() for failures (Petr Písař)
* xqmstats: Fix a file descriptor leak in main() (Petr Písař)
* quotackeck: Fix a directory descriptor leak in scan_dir() (Petr Písař)
* convertquota: Fix a file descriptor leak in convert_endian() (Petr Písař)
* quota(1): Distinguish between none quota limits and no allocated resources (Petr Písař)
* Merge warnquota CC handling fix. (Jan Kara)
* Fix warnquota Cc field being set incorrectly when using LDAP (Santiago Castro)
* Track COPYING file in git. (Jan Kara)
* Add quota_nld to .gitignore (Jan Kara)
* rpc: Fix wrong limit for space usage (Jan Kara)
* Listen on a TCP socket (Steve Dickson)
* Add HPE XFS support (Tinguely, Mark)
* warnquota: Fix comparing user name to non-null-terminated utmp.ut_user (Petr Písař)
* warnquota: Check snprintf() for overflows (Petr Písař)
* Add $(TIRPC_CFLAGS) globally to CFLAGS for RPC support, needed for libc 2.26 (Andreas K. Hüttel)
* Install rquota(3) only if RPC is enabled (Petr Písař)
* quotacheck: Fail check if quota file magic is invalid (Jan Kara)
* quotacheck: Report error when caching of quota file fails (Jan Kara)
* quotacheck: Avoid question in non-interactive mode (Jan Kara)
* repquota: Fix output when user -2 exists (Jan Kara)
* Implement manual choice of output unit (Jan Kara)
* f2fs: support f2fs's quota sysfile (Jaegeuk Kim)

Changes in quota-tools from 4.03 to 4.04
* Use TI-RPC for RPC client communication (Jan Kara)
* rpc.rquotad: IPv6 support (Jan Kara)
* rpc: Factor out getting of port number (Jan Kara)
* Add support for ipv6 to good_client (Jan Kara)
* quotacheck: Deallocate memory after direct scanning (Petr Písař)
* quota: Return non-zero exit code when getting quota fails (Ian Chard)
* quotaops: check return code of ftruncate and lseek calls (Dmitry V. Levin)
* Use configure macros instead of hardcoded defaults (Dmitry V. Levin)
* quotaops: check return code of fgets calls (Dmitry V. Levin)
* quotaops: check setgid/setuid return code (Dmitry V. Levin)
* quotacheck: fix ask_yn UB when fgets returns NULL (Dmitry V. Levin)
* quotacheck: change to the directory before opening it, not after (Dmitry V. Levin)
* Do not install quota_nld.8 when quota_nld is not installed (Dmitry V. Levin)
* quota-tools: add missing newline to die() messages (Theodore Ts'o)
* quotacheck: Use direct scanning also for ext4 (Jan Kara)
* Improve detection of XFS quota interface (Jan Kara)
* quotaon: Improve reporting of quota state (Jan Kara)
* quota: fix coredump if projid file does not exist (Wang Shilong)
* quotacheck: Fix buggy error check of read(2) (Jan Kara)
* repquota: use the same whitespace for quotaio_meta as quotaio_v2 (Theodore Ts'o)
* repquota: Add project quota support (Jan Kara)
* edquota: Project quota support (Jan Kara)
* setquota: Project quota support (Jan Kara)
* quota: Add project quota support (Jan Kara)
* quotasync: Add project quota support (Jan Kara)
* quotaon: Add project quota support (Jan Kara)
* Add support for project quota into generic code (Jan Kara)
* Fix warnings due to missing stdlib.h (Jan Kara)
* Support systems which do not have nl_langinfo() (Theodore Ts'o)
* Provide fallback definitions for MNTTYPE_NFS and MNTTYPE_NOAUTO (Theodore Ts'o)
* Use NGROUPS_MAX instead of NGROUPS (Theodore Ts'o)
* Allow building on systems that do not have rpc header files (Theodore Ts'o)
* Update the gitignore file to ignore additional generated files (Theodore Ts'o)
* Don't build rpc.rquotad when --disable-rpc was requested. (Lars Wendler)
* Add support for scanning using Q_XGETNEXTQUOTA (Jan Kara)
* Scan dquots using Q_GETNEXTQUOTA (Jan Kara)
* Don't link all binaries with ldap library (Jan Kara)
* repquota: -F option takes an arg (Eric Sandeen)
* Set -fPIC and -pie as default params when building (Tomáš Chvátal)
* Do not accidentaly override commandline passed CFLAGS. (Tomáš Chvátal)
* Respect the docdir declaration and do not override. (Tomáš Chvátal)
* Print explicitely disabled options properly (Jan Kara)
* Distribute ldap-scripts directory too (Tomáš Chvátal)
* Fix build with disabled ldap (Jan Kara)
* Fix whitespace in configure.ac (Tomáš Chvátal)

Changes in quota-tools from 4.02 to 4.03
* Revive Changelog (Jan Kara)
* Update release script to work with new build system (Jan Kara)
* Make new build system use original versioning scheme (Jan Kara)
* Automatic update of translation files (Jan Kara)
* warnquota: Added StartTLS support (Ian Allison)
* quota_nld: multicast project quota warnings via dbus (Masatake YAMATO)
* Fix building of rquota.[ch] files (Jan Kara)
* edquota: fix mtime check (Konstantin Khlebnikov)
* Makefile.am: remove quotaoff symlink before install (Konstantin Khlebnikov)
* warnquota: Fix some leaks in LDAP lookup code (Jan Kara)
* warnquota: Don't CC empty names (Jan Kara)
* Fix typos in German translation (Michael Meskes)
* quota_nld: Fixup build (Jan Kara)
* edquota: Do not unnecessarily overwrite limits/times (Jan Kara)
* edquota: Fix error handling (Jan Kara)
* edquota: Reopen the file with new grace times (Jan Kara)
* edquota: Unify code for editing grace times and quota limits (Jan Kara)
* Fix the po generating by accidentaly ignoring POTFILES.in file (Tomáš Chvátal)
* Install example file to docs. (Tomáš Chvátal)
* Ignore generated tarballs. (Tomáš Chvátal)
* config.h.in is autogenerated. (Tomáš Chvátal)
* Use just one define to determine if using ext2direct (Tomáš Chvátal)
* Stop using deprecated ldap_perror (Tomáš Chvátal)
* Refresh and fix the po files. (Tomáš Chvátal)

quota.c
/*
 * Basic utility for reporting quotas
 */

#include "config.h"

/*
 * Disk quota reporting program.
 */
#include <sys/types.h>
#include <sys/param.h>
#include <getopt.h>
#include <stdio.h>
#include <stdlib.h>
#include <pwd.h>
#include <grp.h>
#include <time.h>
#include <errno.h>
#include <string.h>
#include <unistd.h>
#include <limits.h>
#include <ctype.h>
#ifdef RPC
#include <rpc/rpc.h>
#include "rquota.h"
#endif

#include "quota.h"
#include "quotaops.h"
#include "quotasys.h"
#include "pot.h"
#include "common.h"

#define FL_QUIET 1
#define FL_VERBOSE 2
#define FL_USER 4
#define FL_GROUP 8
#define FL_LOCALONLY 32
#define FL_QUIETREFUSE 64
#define FL_NOAUTOFS 128
#define FL_NOWRAP 256
#define FL_FSLIST 512
#define FL_NUMNAMES 1024
#define FL_NFSALL 2048
#define FL_RAWGRACE 4096
#define FL_NO_MIXED_PATHS 8192
#define FL_SHOW_MNTPOINT 16384
#define FL_SHOW_DEVICE 32768
#define FL_PROJECT 65536

static int flags, fmt = -1;
static enum s2s_unit spaceunit = S2S_NONE, inodeunit = S2S_NONE;
char *progname;

static void usage(void)
{
	errstr( "%s%s%s%s%s%s",
		_("Usage: quota [-guPqvswim] [-l | [-Q | -A]] [-F quotaformat]\n"),
		_("\tquota [-qvswim] [-l | [-Q | -A]] [-F quotaformat] -u username ...\n"),
		_("\tquota [-qvswim] [-l | [-Q | -A]] [-F quotaformat] -g groupname ...\n"),
		_("\tquota [-qvswim] [-l | [-Q | -A]] [-F quotaformat] -P projectname ...\n"),
		_("\tquota [-qvswugPQm] [-F quotaformat] -f filesystem ...\n"),
		_("\n\
-u, --user                    display quota for user\n\
-g, --group                   display quota for group\n\
-P, --project                 display quota for project\n\
-q, --quiet                   print more terse message\n\
-v, --verbose                 print more verbose message\n\
-s, --human-readable[=units]  display numbers in human friendly units (MB, GB,\n\
                              ...). Units can be also specified explicitely by\n\
                              an optional argument in format [kgt],[kgt] where\n\
                              the first character specifies space units and the\n\
                              second character specifies inode units\n\
    --always-resolve          always try to translate name to id, even if it is\n\
			      composed of only digits\n\
-w, --no-wrap                 do not wrap long lines\n\
-p, --raw-grace               print grace time in seconds since epoch\n\
-l, --local-only              do not query NFS filesystems\n\
-Q, --quiet-refuse            do not print error message when NFS server does\n\
                              not respond\n\
-i, --no-autofs               do not query autofs mountpoints\n\
-F, --format=formatname       display quota of a specific format\n\
-f, --filesystem-list         display quota information only for given\n\
                              filesystems\n\
    --filesystem=path         display quota information only for given\n\
                              filesystem, remaining command line arguments\n\
                              are still treated as user/group/project names\n\
-A, --all-nfs                 display quota for all NFS mountpoints\n\
-m, --no-mixed-pathnames      trim leading slashes from NFSv4 mountpoints\n\
    --show-mntpoint           show mount point of the file system in output\n\
    --hide-device             do not show file system device in output\n\
-h, --help                    display this help message and exit\n\
-V, --version                 display version information and exit\n\n"));
	fprintf(stderr, _("Bugs to: %s\n"), PACKAGE_BUGREPORT);
	exit(1);
}

static void heading(int type, qid_t id, char *name, char *tag)
{
	char *spacehdr;

	if (spaceunit != S2S_NONE)
		spacehdr = _("space");
	else
		spacehdr = _("blocks");

	printf(_("Disk quotas for %s %s (%cid %u): %s\n"), _(type2name(type)),
	       name, *type2name(type), (uint) id, tag);
	if (!(flags & FL_QUIET) && !tag[0]) {
		printf("%15s%8s %7s%8s%8s%8s %7s%8s%8s\n", _("Filesystem"),
		       spacehdr, _("quota"), _("limit"), _("grace"),
		       _("files"), _("quota"), _("limit"), _("grace"));
	}
}

static void print_fs_location(struct dquot *q)
{
	struct quota_handle *h = q->dq_h;

	if (flags & FL_QUIET) {
		if (flags & FL_SHOW_DEVICE)
			printf(" %s", h->qh_quotadev);
		if (flags & FL_SHOW_MNTPOINT)
			printf(" %s", h->qh_dir);
		putchar('\n');
	} else {
		int wrap = 0;

		if (flags & FL_SHOW_DEVICE && flags & FL_SHOW_MNTPOINT &&
		    !(flags & FL_NOWRAP))
			wrap = 1;
		else if (flags & FL_SHOW_DEVICE && strlen(h->qh_quotadev) > 15 &&
		    !(flags & FL_NOWRAP))
			wrap = 1;
		else if (flags & FL_SHOW_MNTPOINT && strlen(h->qh_dir) > 15 &&
		    !(flags & FL_NOWRAP))
			wrap = 1;
		
		if (flags & FL_SHOW_DEVICE) {
			if (wrap || flags & FL_SHOW_MNTPOINT)
				printf("%s", h->qh_quotadev);
			else
				printf("%15s", h->qh_quotadev);
		}
		if (flags & FL_SHOW_MNTPOINT) {
			if (flags & FL_SHOW_DEVICE)
				putchar(' ');
			if (wrap || flags & FL_SHOW_DEVICE)
				printf("%s", h->qh_dir);
			else
				printf("%15s", h->qh_dir);
		}
		if (wrap)
			printf("\n%15s", "");
	}
}

static int showquotas(int type, qid_t id, int mntcnt, char **mnt)
{
	struct dquot *qlist, *q;
	char *msgi, *msgb;
	char timebuf[MAXTIMELEN];
	char name[MAXNAMELEN];
	struct quota_handle **handles;
	int lines = 0, bover, iover, over, unlimited;
	time_t now;

	time(&now);
	id2name(id, type, name);
	handles = create_handle_list(mntcnt, mnt, type, fmt,
		IOI_READONLY | ((flags & FL_NO_MIXED_PATHS) ? 0 : IOI_NFS_MIXED_PATHS),
		((flags & FL_NOAUTOFS) ? MS_NO_AUTOFS : 0)
		| ((flags & FL_LOCALONLY) ? MS_LOCALONLY : 0)
		| ((flags & FL_NFSALL) ? MS_NFS_ALL : 0));
	qlist = getprivs(id, handles, !mntcnt || (flags & FL_QUIETREFUSE));
	if (!qlist) {
		over = 1;
		goto out_handles;
	}
	over = 0;
	unlimited = 1;
	for (q = qlist; q; q = q->dq_next) {
		bover = iover = 0;
		if (!q->dq_dqb.dqb_isoftlimit && !q->dq_dqb.dqb_ihardlimit
		    && !q->dq_dqb.dqb_bsoftlimit && !q->dq_dqb.dqb_bhardlimit) {
			if (!(flags & FL_VERBOSE))
				continue;
		} else {
			unlimited = 0;
		}
		msgi = NULL;
		if (q->dq_dqb.dqb_ihardlimit && q->dq_dqb.dqb_curinodes >= q->dq_dqb.dqb_ihardlimit) {
			msgi = _("File limit reached on");
			iover = 1;
		}
		else if (q->dq_dqb.dqb_isoftlimit
			 && q->dq_dqb.dqb_curinodes > q->dq_dqb.dqb_isoftlimit) {
			if (q->dq_dqb.dqb_itime > now) {
				msgi = _("In file grace period on");
				iover = 2;
			}
			else {
				msgi = _("Over file quota on");
				iover = 3;
			}
		}
		msgb = NULL;
		if (q->dq_dqb.dqb_bhardlimit && toqb(q->dq_dqb.dqb_curspace) >= q->dq_dqb.dqb_bhardlimit) {
				msgb = _("Block limit reached on");
				bover = 1;
		}
		else if (q->dq_dqb.dqb_bsoftlimit
			 && toqb(q->dq_dqb.dqb_curspace) > q->dq_dqb.dqb_bsoftlimit) {
			if (q->dq_dqb.dqb_btime > now) {
				msgb = _("In block grace period on");
				bover = 2;
			}
			else {
				msgb = _("Over block quota on");
				bover = 3;
			}
		}
		over |= bover | iover;
		if (flags & FL_QUIET) {
			if ((msgi || msgb) && !lines++)
				heading(type, id, name, "");
			if (msgi) {
				printf("\t%s", msgi);
				print_fs_location(q);
			}
			if (msgb) {
				printf("\t%s", msgb);
				print_fs_location(q);
			}
			continue;
		}
		if ((flags & FL_VERBOSE) || q->dq_dqb.dqb_curspace || q->dq_dqb.dqb_curinodes) {
			char numbuf[3][MAXNUMLEN];

			if (!lines++)
				heading(type, id, name, "");
			print_fs_location(q);
			if (!(flags & FL_RAWGRACE)) {
				if (bover)
					difftime2str(q->dq_dqb.dqb_btime, timebuf);
				else
					timebuf[0] = 0;
			}
			else {
				if (bover)
					sprintf(timebuf, "%llu", (long long unsigned int)q->dq_dqb.dqb_btime);
				else
					strcpy(timebuf, "0");
			}
			space2str(toqb(q->dq_dqb.dqb_curspace), numbuf[0], spaceunit);
			space2str(q->dq_dqb.dqb_bsoftlimit, numbuf[1], spaceunit);
			space2str(q->dq_dqb.dqb_bhardlimit, numbuf[2], spaceunit);
			printf(" %7s%c %6s %7s %7s", numbuf[0], bover ? '*' : ' ', numbuf[1],
			       numbuf[2], timebuf);

			if (!(flags & FL_RAWGRACE)) {
				if (iover)
					difftime2str(q->dq_dqb.dqb_itime, timebuf);
				else
					timebuf[0] = 0;
			}
			else {
				if (iover)
					sprintf(timebuf, "%llu", (long long unsigned int)q->dq_dqb.dqb_itime);
				else
					strcpy(timebuf, "0");
			}
			number2str(q->dq_dqb.dqb_curinodes, numbuf[0], inodeunit);
			number2str(q->dq_dqb.dqb_isoftlimit, numbuf[1], inodeunit);
			number2str(q->dq_dqb.dqb_ihardlimit, numbuf[2], inodeunit);
			printf(" %7s%c %6s %7s %7s\n", numbuf[0], iover ? '*' : ' ', numbuf[1],
			       numbuf[2], timebuf);
			continue;
		}
	}
	if (!(flags & FL_QUIET) && !lines && qlist)
		heading(type, id, name, unlimited ? _("none") : _("no limited resources used"));
	freeprivs(qlist);
out_handles:
	dispose_handle_list(handles);
	return over > 0 ? 1 : 0;
}

int main(int argc, char **argv)
{
	int ngroups;
	gid_t gidset[NGROUPS_MAX], *gidsetp;
	char **fsnames = NULL;
	int fscount = 0;
	int i, ret, type = 0;
	struct option long_opts[] = {
		{ "help", 0, NULL, 'h' },
		{ "version", 0, NULL, 'V' },
		{ "user", 0, NULL, 'u' },
		{ "group", 0, NULL, 'g' },
		{ "project", 0, NULL, 'P' },
		{ "quiet", 0, NULL, 'q' },
		{ "verbose", 0, NULL, 'v' },
		{ "human-readable", 2, NULL, 's' },
		{ "always-resolve", 0, NULL, 256 },
		{ "raw-grace", 0, NULL, 'p' },
		{ "local-only", 0, NULL, 'l' },
		{ "no-autofs", 0, NULL, 'i' },
		{ "quiet-refuse", 0, NULL, 'Q' },
		{ "format", 1, NULL, 'F' },
		{ "no-wrap", 0, NULL, 'w' },
		{ "filesystem-list", 0, NULL, 'f' },
		{ "all-nfs", 0, NULL, 'A' },
		{ "no-mixed-pathnames", 0, NULL, 'm' },
		{ "show-mntpoint", 0, NULL, 257 },
		{ "hide-device", 0, NULL, 258 },
		{ "filesystem", 1, NULL, 259 },
		{ NULL, 0, NULL, 0 }
	};

	gettexton();
	progname = basename(argv[0]);

	flags |= FL_SHOW_DEVICE;
	while ((ret = getopt_long(argc, argv, "hguqvs::PVliQF:wfApm", long_opts, NULL)) != -1) {
		switch (ret) {
		  case 'g':
			  flags |= FL_GROUP;
			  break;
		  case 'u':
			  flags |= FL_USER;
			  break;
		  case 'P':
			  flags |= FL_PROJECT;
			  break;
		  case 'q':
			  flags |= FL_QUIET;
			  break;
		  case 'v':
			  flags |= FL_VERBOSE;
			  break;
		  case 'F':
			  if ((fmt = name2fmt(optarg)) == QF_ERROR)	/* Error? */
				  exit(1);
			  break;
		  case 's':
			  inodeunit = spaceunit = S2S_AUTO;
			  if (optarg) {
				if (unitopt2unit(optarg, &spaceunit, &inodeunit) < 0)
					die(1, _("Bad output format units for human readable output: %s\n"), optarg);
			  }
			  break;
		  case 'p':
			  flags |= FL_RAWGRACE;
			  break;
		  case 256:
			  flags |= FL_NUMNAMES;
			  break;
		  case 'l':
			  flags |= FL_LOCALONLY;
			  break;
		  case 'Q':
			  flags |= FL_QUIETREFUSE;
			  break;
		  case 'i':
			  flags |= FL_NOAUTOFS;
			  break;
		  case 'w':
			  flags |= FL_NOWRAP;
			  break;
		  case 'f':
			  flags |= FL_FSLIST;
			  break;
		  case 'A':
			  flags |= FL_NFSALL;
			  break;
		  case 'm':
			  flags |= FL_NO_MIXED_PATHS;
			  break;
		  case 257:
			  flags |= FL_SHOW_MNTPOINT;
			  break;
		  case 258:
			  flags &= ~FL_SHOW_DEVICE;
			  break;
		  case 259:
			  fscount++;
			  fsnames = reallocarray(fsnames, fscount, sizeof(char *));
			  if (!fsnames)
				die(1, _("Not enough memory for filesystem names"));
			  fsnames[fscount - 1] = optarg;
			  break;
		  case 'V':
			  version();
			  exit(0);
		  case 'h':
		  default:
			  usage();
		}
	}
	argc -= optind;
	argv += optind;

	if (!(flags & FL_USER) && !(flags & FL_GROUP) && !(flags & FL_PROJECT))
		flags |= FL_USER;
	if (((flags & FL_FSLIST) || fscount) && flags & (FL_LOCALONLY | FL_NOAUTOFS))
		errstr(_("Warning: Ignoring -%c when filesystem list specified.\n"), flags & FL_LOCALONLY ? 'l' : 'i');
	if (fscount && flags & FL_FSLIST)
		die(1, "Cannot use both --filesystem and -f");

	init_kernel_interface();

	ret = 0;
	if (argc == 0 || flags & FL_FSLIST) {
		if (flags & FL_FSLIST && argc == 0)
			die(1, _("No filesystem specified.\n"));
		if (flags & FL_USER)
			ret |= showquotas(USRQUOTA, getuid(), argc, argv);
		if (flags & FL_GROUP) {
			ngroups = sysconf(_SC_NGROUPS_MAX);
			if (ngroups > NGROUPS_MAX) {
				gidsetp = malloc(ngroups * sizeof(gid_t));
				if (!gidsetp)
					die(1, _("Gid set allocation (%d): %s\n"), ngroups, strerror(errno));
			} else {
				gidsetp = &gidset[0];
			}
			ngroups = getgroups(ngroups, gidsetp);
			if (ngroups < 0)
				die(1, _("getgroups(): %s\n"), strerror(errno));
			for (i = 0; i < ngroups; i++)
				ret |= showquotas(GRPQUOTA, gidsetp[i], argc, argv);
		}
		if (flags & FL_PROJECT)
			die(1, _("Project reports not supported without project name\n"));
		exit(ret);
	}

	if (flags & FL_USER)
		type++;
	if (flags & FL_GROUP)
		type++;
	if (flags & FL_PROJECT)
		type++;
	if (type > 1)
		usage();

	if (flags & FL_USER)
		for (; argc > 0; argc--, argv++)
			ret |= showquotas(USRQUOTA, user2uid(*argv, !!(flags & FL_NUMNAMES), NULL), fscount, fsnames);
	else if (flags & FL_GROUP)
		for (; argc > 0; argc--, argv++)
			ret |= showquotas(GRPQUOTA, group2gid(*argv, !!(flags & FL_NUMNAMES), NULL), fscount, fsnames);
	else if (flags & FL_PROJECT)
		for (; argc > 0; argc--, argv++)
			ret |= showquotas(PRJQUOTA, project2pid(*argv, !!(flags & FL_NUMNAMES), NULL), fscount, fsnames);
	return ret;
}
